### PR TITLE
Add setuptools

### PR DESCRIPTION
**Pull request by jakirkham**
Closes https://github.com/conda-forge/staged-recipes/pull/551

Takes over @msarahan's great work of packaging `setuptools`.

### CODE REVIEW

**Comment by jakirkham** [comment]
FYI. Removing `setuptools`, `pip`, et al as dependencies appears to be crucial for moving this forward. This built fine in the docker container locally with this change.

File: recipes/setuptools/setup.patch [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+@@ -145,12 +145,8 @@ setup_params = dict(
+         Topic :: Utilities
+         """).strip().splitlines(),
+     extras_require={
+-        "ssl:sys_platform=='win32'": "wincertstore==0.2",
+-        "certs": "certifi==2015.11.20",
+     },
+     dependency_links=[
+-        'https://pypi.python.org/packages/source/c/certifi/certifi-2015.11.20.tar.gz#md5=25134646672c695c1ff1593c2dd75d08',
+-        'https://pypi.python.org/packages/source/w/wincertstore/wincertstore-0.2.zip#md5=ae728f2f007185648d0c7a8679b361e2',
```
Comment (jakirkham) [inline recipes/setuptools/setup.patch]: Should we add these as run requirements, @msarahan?

**Comment by jakirkham** [comment]
This passes. 😄 It is also able to import `pkg_resources` fine, which is a good sign. There is still more to do here though.

File: recipes/setuptools/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  commands:
+    - easy_install --help
+  imports:
+    - setuptools
+    - easy_install
+    - pkg_resources
+
+about:
+  home: https://pypi.python.org/pypi/setuptools
+  license: PSF or ZPL
```
Comment (jakirkham) [inline recipes/setuptools/meta.yaml]: So it says on PyPI that this MIT (via a [classifier]( https://pypi.python.org/pypi/setuptools/22.0.5#code-of-conduct )).

Comment (jakirkham) [inline recipes/setuptools/meta.yaml]: That seems to be all the information I can find ATM.

File: recipes/setuptools/setup.patch [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+@@ -145,12 +145,8 @@ setup_params = dict(
+         Topic :: Utilities
+         """).strip().splitlines(),
+     extras_require={
+-        "ssl:sys_platform=='win32'": "wincertstore==0.2",
+-        "certs": "certifi==2016.2.28",
+     },
+     dependency_links=[
+-        'https://pypi.python.org/packages/source/c/certifi/certifi-2016.2.28.tar.gz#md5=5d672aa766e1f773c75cfeccd02d3650',
+-        'https://pypi.python.org/packages/source/w/wincertstore/wincertstore-0.2.zip#md5=ae728f2f007185648d0c7a8679b361e2',
```
Comment (jakirkham) [inline recipes/setuptools/setup.patch]: Should we add these as runtime dependencies?

**Comment by jakirkham** [comment]
Alright I have a few questions, but this is open to feedback.

File: recipes/setuptools/setup.patch (hunk 3 above)
Comment (msarahan) [inline recipes/setuptools/setup.patch]: If we can do so without creating a troublesome circular dependency, then sure.

Comment (jakirkham) [inline recipes/setuptools/setup.patch]: Good point. `certifi` requires [`setuptools`]( https://github.com/certifi/python-certifi/blob/2016.02.28/setup.py#L7 ). `wincertstore` and can live [without `setuptools`]( https://bitbucket.org/tiran/wincertstore/src/7eb3f9d9a5af79d3c284ba04fed9aee613f16b1d/setup.py?at=0.2&fileviewer=file-view-default#setup.py-4:7 ). Though I'm a bit confused why both don't use `distutils` outright as they don't seem to use any special `setuptools` features.

File: recipes/setuptools/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,50 @@
+{% set version = "22.0.5" %}
```
Comment (jakirkham) [inline recipes/setuptools/meta.yaml]: LOL they just bumped to 23.0.0. I wonder if we will ever release the current version. 😆

Comment (jakirkham) [inline recipes/setuptools/meta.yaml]: Fortunately the patches didn't need updating.

File: recipes/setuptools/setup.patch (hunk 3 above)
Comment (jakirkham) [inline recipes/setuptools/setup.patch]: This constraint has been relaxed in `master` of `certifi`. See PR ( https://github.com/certifi/python-certifi/pull/43 ) and issue ( https://github.com/certifi/python-certifi/pull/42 ) for details.

File: recipes/setuptools/meta.yaml (hunk 2 above)
Comment (jakirkham) [inline recipes/setuptools/meta.yaml]: Added this issue ( https://github.com/pypa/setuptools/issues/612 ) for license clarification. Let's see what they say.

Comment (jakirkham) [inline recipes/setuptools/meta.yaml]: I heard back from upstream and they confirmed this MIT. Updating respectively.

File: recipes/setuptools/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    - easy_install --help
+  imports:
+    - setuptools
+    - easy_install
+    - pkg_resources
+
+about:
+  home: https://pypi.python.org/pypi/setuptools
+  license: MIT
+  license_family: MIT
```
Comment (jakirkham) [inline recipes/setuptools/meta.yaml]: See this issue ( https://github.com/pypa/setuptools/issues/612 ) for details.

**Comment by ilanschnell** [comment]
When we build packages for Anaconda, we always set (in `~/.condarc`):
```
add_pip_as_python_dependency: False
```

**Comment by jakirkham** [comment]
Yep, we recently decided to start doing that here too as it is too confusing to determine what is going on otherwise. Though we still need to deploy this to the feedstocks, it is definitely the case at staged-recipes.

We have refactored out some of the configuration between the two as well. So, it isn't immediately obvious that we are doing this. We have a package with configuration scripts. [Here]( https://github.com/conda-forge/conda-forge-build-setup-feedstock/blob/646cd541768932c480724b6d41801bd973b434f6/recipe/run_conda_forge_build_setup_linux#L10 ) is one of the scripts used on Linux where we set this configuration.

**Comment by jjhelmus** [comment]
LGTM. Would very much like to see this available in conda-forge.

I think is is fine to not include `certifi` as a run requirement, `setuptools` works without `certifi` and Continuum's `setuptools` package does not have it as a run dependency.
I also do not think `wincertstore` is needed as run dependencies, from my inspection this is not included in the win-32 `setuptools` package in the defaults channel.

**Comment by jakirkham** [comment]
Me too. It's not merge ready though as we need to disable `setuptools` in the feedstocks. I suppose we could add a hacky workaround for the `setuptools` feedstock if we really want to merge this ASAP.

**Comment by jjhelmus** [comment]
Would merging conda-forge/conda-smithy#196 and making a release of conda-smithy fix the feedstocks?

**Comment by jakirkham** [comment]
Yes, it would. I think we should be ok to try that. Though I haven't heard back from anyone so have left it alone.

**Comment by jakirkham** [comment]
Should we remove it from staged-recipes manually too?

**Comment by pelson** [comment]
This ended up going through eventually. This was 100% down to TravisCI API flake 😢  6315e47ddaeb85b5e74670a5de89fccff9ce5fd0

**Comment by jakirkham** [comment]
Yeah, I noticed that this morning.

I really hope Travis Support gives us something to run with in terms of potential fixes or workarounds.

The best I can think of is that client name is overused and that somehow has an effect. I suppose it can't hurt to change it to something like `CondaForge/1.0.0` or something. IDK. Though I think there must be something more to it.

File: recipes/setuptools/setup.patch (hunk 3 above)
Comment (jakirkham) [inline recipes/setuptools/setup.patch]: These were added in PR ( https://github.com/conda-forge/setuptools-feedstock/pull/1 ) and the patch was dropped in PR ( https://github.com/conda-forge/setuptools-feedstock/pull/7 ).

**Pull request merged**